### PR TITLE
Respect limit during join order

### PR DESCRIPTION
**Pull request by Tmonster**
while iterating through the relations, check to see if a limit is encountered. If so modify the estimated cardinality of the added relation accordingly. 

Right now this only includes the case where the limit val is constant, but this logic can most likely be expanded to include cases where it is a percentage as well. 

No attention is paid to the offset value. My thinking is that the limit value is what is important to the optimizer when choosing a build side and estimating the cardinality of a join. If someone thinks otherwise please let me know

### CODE REVIEW

File: test/optimizer/joins/get_cardinality_from_limit.test [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+statement ok
+create table t_left as select (random() * 10000000)::INT a from range(40000);
+
+statement ok
+create table t_right as select range b from range(10000000);
+
+statement ok
+pragma explain_output='optimized_only'
+
+
```
Comment (Tmonster) [inline test/optimizer/joins/get_cardinality_from_limit.test:14]: This needs a REGEX check, but the plans look very weird, so I don't know exactly what I want here

File: src/optimizer/join_order/relation_manager.cpp [hunk 2]
```diff
@@ -133,11 +133,22 @@ static bool HasNonReorderableChild(LogicalOperator &op) {
...
+		stats.cardinality = MinValue(limit.limit_val.GetConstantValue(), stats.cardinality);
+	}
+}
+
 bool RelationManager::ExtractJoinRelations(LogicalOperator &input_op,
                                            vector<reference<LogicalOperator>> &filter_operators,
                                            optional_ptr<LogicalOperator> parent) {
 	LogicalOperator *op = &input_op;
 	vector<reference<LogicalOperator>> datasource_filters;
+	LogicalOperator *limit_op = nullptr;
```
Comment (lnkuiper) [inline src/optimizer/join_order/relation_manager.cpp:151]: Can we make this an `optional_ptr`?

File: test/optimizer/joins/get_cardinality_from_limit.test (hunk 1 above)
Comment (lnkuiper) [reply test/optimizer/joins/get_cardinality_from_limit.test:14]: If the plan looks like this:
```
┌───────────────────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             i             │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         HASH_JOIN         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           INNER           │                             
│           i = i           │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ├──────────────┐              
│        Build Min: 0       │              │              
│        Build Max: 9       │              │              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │              
│           EC: 10          │              │              
└─────────────┬─────────────┘              │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││           LIMIT           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │
│             t0            ││                           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │
│             i             ││                           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │
│ Filters: i<=9 AND i IS NOT││                           │
│            NULL           ││                           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││                           │
│           EC: 11          ││                           │
└───────────────────────────┘└─────────────┬─────────────┘                             
                             ┌─────────────┴─────────────┐
                             │         SEQ_SCAN          │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │             t1            │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │             i             │
                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
                             │           EC: 10          │
                             └───────────────────────────┘
```
You can probably do something like `.*SEQ_SCAN.*LIMIT.*` to check that the limit is on the rhs (build side).

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Looks good! I just have two comments:

File: src/optimizer/join_order/relation_manager.cpp [hunk 3]
```diff
@@ -133,11 +133,22 @@ static bool HasNonReorderableChild(LogicalOperator &op) {
 	return tmp->children.empty();
 }
 
+static void ModifyStatsIfLimit(LogicalOperator *limit_op, RelationStats &stats) {
```
Comment (Mytherin) [inline src/optimizer/join_order/relation_manager.cpp:136]: nit: Should this be a raw pointer?

**Review by Mytherin — COMMENTED** [review]
Thanks! LGTM - one minor nit

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**